src/sieve/sampler_basic.cpp

```diff
@@ -0,0 +1,172 @@
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+#include "sampler_basic.h"
+
+
+/**************************
+ *  class KleinSampler
+ **************************/
+
+
+template<class ZT, class F>
+KleinSampler<ZT, F>::KleinSampler (ZZ_mat<ZT> &B, bool ver)
+{
+  /* set dimensions */
+  b = B;
+  nr = b.getRows();
+  nc = b.getCols();
+  //t = log(nr);
+  t = 2;
+  logn2 = log(nr)*log(nr);
+
+  /* gso, flag 1 to have g matrix valid */
+  pGSO = new MatGSO<Z_NR<ZT>, F> (b, u, uInv, 1);
+
+  pGSO->updateGSO();
+  mu = pGSO->getMuMatrix();
+  r = pGSO->getRMatrix();
+  g = pGSO->getGMatrix();
+  
+  /* compute variances for sampling */
+  maxbistar2 = pGSO->getMaxBstar();
+  s2.mul_d (maxbistar2, logn2, GMP_RNDN);
+  s_prime = new NumVect<F> (nr);
+  F tmp;
+  for (int i = 0; i < nr; i++) {
+    tmp.set_z(g(i,i));
+    ((*s_prime)[i]).div(maxbistar2, tmp);
+    ((*s_prime)[i]).sqrt((*s_prime)[i], GMP_RNDN);
+  }
+
+  /* verbose */
+  set_verbose(ver);
+  print_param();
+  gmp_randinit_default(state);
```

Comment:
Why does the sampler maintain its own state? Wouldn't it be better to use the common state for reproducibility?